File: src/types/DerivedCharacteristic.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { AttributeIdentifier } from "./_Identifier.ts"
import { MathOperation } from "./_MathExpression.ts"
import { DerivedCharacteristicPrerequisites } from "./_Prerequisite.js"
import { src } from "./source/_PublicationRef.js"

export const DerivedCharacteristic = DB.Entity(import.meta.url, {
  name: "DerivedCharacteristic",
  namePlural: "DerivedCharacteristics",
  type: () =>
    DB.Object({
      position: DB.Required({
        comment:
          "The position of the derived characteristic in lists. This has to be a unique value.",
        type: DB.Integer({ minimum: 0 }),
      }),
      calculation: DB.Required({
        comment: "Instructions for calculating the derived characteristic’s value.",
        type: DB.Object({
          base: DB.Required({
            comment: "The base value for the calculation.",
            type: DB.IncludeIdentifier(BaseCalculation),
          }),
        }),
      }),
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(DerivedCharacteristicPrerequisites),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "DerivedCharacteristic",
        DB.Object({
          name: DB.Required({
            comment: "The derived characteristic’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          abbreviation: DB.Required({
            comment: "The derived characteristic’s abbreviation.",
            type: DB.String({ minLength: 1 }),
          }),
          calculation: DB.Optional({
            comment: "Possible calculation strings for the final value.",
            type: DB.IncludeIdentifier(CalculationTranslation),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  sortOrder: {
    keyPath: "position",
  },
  uniqueConstraints: [
    {
      keyPath: "position",
    },
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "abbreviation",
    },
  ],
})

const RaceBaseCalculationValue = DB.Enum(import.meta.url, {
  name: "RaceBaseCalculationValue",
  values: () => ({
    LifePoints: DB.EnumCase({ type: null }),
    Spirit: DB.EnumCase({ type: null }),
    Toughness: DB.EnumCase({ type: null }),
    Movement: DB.EnumCase({ type: null }),
  }),
})

const BaseCalculationValue = DB.Enum(import.meta.url, {
  name: "BaseCalculationValue",
  values: () => ({
    Constant: DB.EnumCase({ type: DB.Integer() }),
    Attribute: DB.EnumCase({ type: AttributeIdentifier() }),
    RaceBaseValue: DB.EnumCase({
      displayName: "Base Value from Race",
      type: DB.IncludeIdentifier(RaceBaseCalculationValue),
    }),
  }),
})

const BaseCalculation = DB.TypeAlias(import.meta.url, {
  name: "BaseCalculation",
  type: () => DB.GenIncludeIdentifier(MathOperation, [DB.IncludeIdentifier(BaseCalculationValue)]),
})

const CalculationTranslation = DB.TypeAlias(import.meta.url, {
  name: "CalculationTranslation",
  type: () =>
    DB.Object({
      default: DB.Required({
        comment: "The default calculation string.",
        type: DB.String({ minLength: 1 }),
      }),
      half_primary: DB.Optional({
        comment: "The calculation string if only half of the primary attribute is used.",
        type: DB.String({ minLength: 1 }),
      }),
      no_primary: DB.Optional({
        comment: "The calculation string if no primary attribute is used.",
        type: DB.String({ minLength: 1 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/DerivedCharacteristic.swift
//
//  DerivedCharacteristic.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

public struct DerivedCharacteristic: LocalizableEntity {
    /// An unique, increasing integer.
    public let id: Int
    
    public let prerequisites: DerivedCharacteristicPrerequisites?
    
    public let src: PublicationRefs
    
    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<DerivedCharacteristicTranslation>

    public init(id: Int, prerequisites: DerivedCharacteristicPrerequisites? = nil, src: PublicationRefs, translations: LocaleMap<DerivedCharacteristicTranslation>) {
        self.id = id
        self.prerequisites = prerequisites
        self.src = src
        self.translations = translations
    }
}

public struct DerivedCharacteristicTranslation: EntitySubtype {
    /// The characteristic's name.
    public let name: String
    
    /// The characteristic's abbreviation.
    public let abbreviation: String
    
    /// Possible calculation strings for the final value.
    public let calculation: CalculationTranslation?

    public init(name: String, abbreviation: String, calculation: CalculationTranslation? = nil) {
        self.name = name
        self.abbreviation = abbreviation
        self.calculation = calculation
    }
}

/// Possible calculation strings for the final value.
public struct CalculationTranslation: EntitySubtype {
    /// The default calculation string.
    public let `default`: String
    
    /// The calculation string if only half of the primary attribute is used.
    public let halfPrimary: String?
    
    /// The calculation string if no primary attribute is used.
    public let noPrimary: String?

    public init(`default`: String, halfPrimary: String? = nil, noPrimary: String? = nil) {
        self.`default` = `default`
        self.halfPrimary = halfPrimary
        self.noPrimary = noPrimary
    }    
    
    private enum CodingKeys: String, CodingKey {
        case `default` = "default"
        case halfPrimary = "half_primary"
        case noPrimary = "no_primary"
    }
}
